### ISSUE TITLE
TimeDistributed Layer Does Not Support Multiple Outputs

### ISSUE THREAD

**Issue by ffent** [tensorflow/tensorflow#40896]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Linux Ubuntu 18.04
- TensorFlow installed from source or binary: No
- TensorFlow version: 2.2.0
- Python version: 3.6.9
- CUDA/cuDNN version: 10.1/7.6.4
- GPU model and memory: Tesla V100 - 16GB

**Describe the current behavior**
The [TimeDistributed layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/TimeDistributed) does not support layers with multiple outputs. This issue is also related to #35824, where the missing support for multiple inputs is mentioned, which leads to the overall missing support for nested structures of inputs and outputs.

**Describe the expected behavior**
The TimeDistributed wrapper should support the wrapping of layers with multiple inputs and multiple outputs.

**Standalone code to reproduce the issue**
```
import tensorflow as tf


class CustomLayer(tf.keras.layers.Layer):
    def __init__(self, name=None, **kwargs):
        super(CustomLayer, self).__init__(name=name, **kwargs)
        self.conv_1 = tf.keras.layers.Conv2D(filters=1, kernel_size=(1, 1))
        self.conv_2 = tf.keras.layers.Conv2D(filters=1, kernel_size=(1, 1))

    def call(self, inputs):
        output_1 = self.conv_1(inputs)
        output_2 = self.conv_2(inputs)

        return output_1, output_2

    def compute_output_shape(self, input_shape):
        output_shape_1 = self.conv_1.compute_output_shape(input_shape)
        output_shape_2 = self.conv_2.compute_output_shape(input_shape)

        return output_shape_1, output_shape_2


if __name__ == "__main__":
    inputs = tf.keras.Input(shape=(None, None, None, 1))

    custom_layer = CustomLayer()
    output_1, output_2 = tf.keras.layers.TimeDistributed(custom_layer)(inputs)

```
**Other info / logs**
```
File "/reproduce/template.py", line 29, in <module>
    output_1, output_2 = tf.keras.layers.TimeDistributed(custom_layer)(inputs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/base_layer.py", line 922, in __call__
    outputs = call_fn(cast_inputs, *args, **kwargs)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/layers/wrappers.py", line 246, in call
    output_shape = self.compute_output_shape(input_shape).as_list()
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/layers/wrappers.py", line 192, in compute_output_shape
    child_output_shape = tensor_shape.TensorShape(child_output_shape)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/tensor_shape.py", line 771, in __init__
    self._dims = [as_dimension(d) for d in dims_iter]
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/tensor_shape.py", line 771, in <listcomp>
    self._dims = [as_dimension(d) for d in dims_iter]
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/tensor_shape.py", line 716, in as_dimension
    return Dimension(value)
  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/tensor_shape.py", line 200, in __init__
    None)
  File "<string>", line 3, in raise_from
TypeError: Dimension value must be integer or None or have an __index__ method, got TensorShape([None, None, None, 1])
```

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.2 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/207c828cbc9741176497c7883729f6a2/35824.ipynb). Thanks!

**Comment by gowthamkpr** [comment]
@ffent This is  a duplicate of this issue #35824, Can we go ahead and close this issue and track it as a single issue? Thanks!

**Comment by ffent** [comment]
@gowthamkpr As mentioned in my issue description, both issues refer to the overall missing support for nested structures within the wrapped layer inputs and outputs. Since both issues are addressed in my recent [pull request](https://github.com/tensorflow/tensorflow/pull/40993), feel free to close one of the issues.

**Comment by gowthamkpr** [comment]
Thanks @ffent I am closing this issue.

**Comment by oawxkw** [comment]
I also observed the following API aliases can cause the same issue in older versions of tensorflow.
Users should be cautious when using them on both CPU and GPU up to tensorflow 2.4.3 (v2.4.2-142-g72bb4c22adb).

- `(tf.keras.layers.TimeDistributed)`, `tf.compat.v1.keras.layers.TimeDistributed`

<details>
  <summary>Code to reproduce the issue in <code>tf.compat.v1.keras.layers.TimeDistributed</code> in older versions</summary>

```python
import tensorflow as tf
print(tf.version.GIT_VERSION, tf.version.VERSION, flush=True)
print(tf.config.list_physical_devices(), flush=True)


class CustomLayer(tf.keras.layers.Layer):
    def __init__(self, name=None, **kwargs):
        super(CustomLayer, self).__init__(name=name, **kwargs)
        self.conv_1 = tf.keras.layers.Conv2D(filters=1, kernel_size=(1, 1))
        self.conv_2 = tf.keras.layers.Conv2D(filters=1, kernel_size=(1, 1))

    def call(self, inputs):
        output_1 = self.conv_1(inputs)
        output_2 = self.conv_2(inputs)

        return output_1, output_2

    def compute_output_shape(self, input_shape):
        output_shape_1 = self.conv_1.compute_output_shape(input_shape)
        output_shape_2 = self.conv_2.compute_output_shape(input_shape)

        return output_shape_1, output_shape_2

try:
    inputs = tf.keras.Input(shape=(None, None, None, 1))

    custom_layer = CustomLayer()
    output_1, output_2 = tf.compat.v1.keras.layers.TimeDistributed(custom_layer)(inputs)
except Exception as e:
    print("Failed! Error:", str(e), flush=True)
else:
    print("Success!", flush=True)
```

On GPU, it does not support layers with multiple outputs, and throws the following error:

```text
v2.4.2-142-g72bb4c22adb 2.4.3
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'), PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]
Failed! Error: Dimension value must be integer or None or have an __index__ method, got value 'TensorShape([None, None, None, 1])' with type '<class 'tensorflow.python.framework.tensor_shape.TensorShape'>'
```

This behavior is also reproducible on my CPU machine:

```text
v2.4.2-142-g72bb4c22adb 2.4.3
[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]
Failed! Error: Dimension value must be integer or None or have an __index__ method, got value 'TensorShape([None, None, None, 1])' with type '<class 'tensorflow.python.framework.tensor_shape.TensorShape'>'
```
</details>

It seems to be fixed in tensorflow 2.5.0 (v2.5.0-rc3-213-ga4dfb8d1a71) and later versions.
